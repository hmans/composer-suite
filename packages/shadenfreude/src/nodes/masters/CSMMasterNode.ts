import { node, nodeFactory, Value, vec3 } from "../.."

export const CSMMasterNode = nodeFactory<{
  position?: Value<"vec3">
  normal?: Value<"vec3">
  pointSize?: Value<"float">
  diffuseColor?: Value<"vec3">
  fragColor?: Value<"vec3">
  emissiveColor?: Value<"vec3">
}>(({ position, normal, pointSize, diffuseColor, fragColor, emissiveColor }) =>
  node({
    name: "Master Node",
    inputs: {
      diffuseColor: vec3(diffuseColor),
      emissiveColor: vec3(emissiveColor),
      position: vec3(position),
      normal: vec3(normal)
    },
    vertex: {
      body: `
        ${pointSize ? "csm_PointSize = pointSize;" : ""}
        ${position ? "csm_Position = position;" : ""}
        ${normal ? "csm_Normal = normal;" : ""}
      `
    },
    fragment: {
      body: `
        ${diffuseColor ? "csm_DiffuseColor.rgb = diffuseColor;" : ""}
        ${fragColor ? "csm_FragColor.rgb = fragColor;" : ""}
        ${emissiveColor ? "csm_EmissiveColor.rgb = emissiveColor;" : ""}
      `
    }
  })
)
