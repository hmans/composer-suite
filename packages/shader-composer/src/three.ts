export * from "@shader-composer/three"
