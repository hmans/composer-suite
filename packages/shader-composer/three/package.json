{
  "main": "dist/shader-composer-three.cjs.js",
  "module": "dist/shader-composer-three.esm.js"
}
