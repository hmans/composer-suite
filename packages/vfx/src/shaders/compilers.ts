import {
  Attribute,
  Attributes,
  Shader,
  Uniform,
  Uniforms,
  VariableQualifier,
  Varying,
  Varyings
} from "./types"

export const formatValue = (value: any) =>
  typeof value === "number" ? value.toFixed(5) : value

function compileVariable(
  qualifier: VariableQualifier,
  name: string,
  variable: Uniform | Attribute | Varying,
  initializeValue = true
) {
  const valueString =
    initializeValue && "value" in variable && variable.value !== undefined
      ? ` = ${formatValue(variable.value)}`
      : ""

  return `${qualifier} ${variable.type} ${name}${valueString};`
}

function compileVariables(
  qualifier: VariableQualifier,
  variables: Uniforms | Attributes | Varyings,
  initializeValue = true
) {
  return Object.entries(variables)
    .map(([name, variable]) =>
      compileVariable(qualifier, name, variable, initializeValue)
    )
    .join("\n")
}

function compileProgram(shader: Shader, what: "fragment" | "vertex") {
  return `
    ${compileVariables("uniform", shader.uniforms, false)}
    ${
      what === "vertex"
        ? compileVariables("attribute", shader.attributes, false)
        : ""
    }
    ${compileVariables("varying", shader.varyings)}
    ${shader[`${what}Header`]}

    void main() {
      ${shader[`${what}Main`]}
    }`
}

export function compileShader(shader: Shader) {
  return {
    uniforms: shader.uniforms,
    vertexShader: compileProgram(shader, "vertex"),
    fragmentShader: compileProgram(shader, "fragment"),
    update: shader.update!
  }
}
