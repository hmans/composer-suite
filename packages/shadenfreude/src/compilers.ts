import { RenderCallback } from "@react-three/fiber"
import { IUniform } from "three"
import { variablesToNodes } from "./factories"
import { formatValue } from "./formatters"
import { isVariable, ShaderNode, Value, Variable } from "./types"

type Program = "vertex" | "fragment"

export function compileVariableValue(value: Value): string {
  return isVariable(value) ? value.name : formatValue(value)
}

export function compileVariable(variable: Variable) {
  const valueString =
    variable.qualifier !== "uniform" && variable.value !== undefined
      ? ` = ${compileVariableValue(variable.value)}`
      : ""

  return `${variable.qualifier ?? ""} ${variable.type} ${
    variable.name
  }${valueString};`
}

function nodeTitle(node: ShaderNode) {
  return `/** Node: ${node.name} **/`
}

function dependencies(node: ShaderNode, deps = new Array<ShaderNode>()) {
  for (const variable of Object.values(node.inputs)) {
    if (isVariable(variable.value)) {
      /* get dependency */
      const dependency = variablesToNodes.get(variable.value)!
      if (!dependency) throw new Error("Dependency not found")

      /* If we haven't seen this dependency yet, invoke the callback */
      if (!deps.includes(dependency)) {
        dependencies(dependency, deps)
      }
    }
  }

  deps.push(node)
  return deps
}

function compileHeader(node: ShaderNode, program: Program) {
  const header = (n: ShaderNode) => `
    ${nodeTitle(n)}

    ${Object.values(n.uniforms)
      .map((variable) => compileVariable({ ...variable, qualifier: "uniform" }))
      .join("\n")}

    ${Object.values(n.varyings)
      .map((variable) => compileVariable({ ...variable, qualifier: "varying" }))
      .join("\n")}

    ${n[program].header ?? ""}
  `

  return dependencies(node)
    .map(header)
    .join("\n\n\n")
}

function compileBody(node: ShaderNode, program: Program) {
  return dependencies(node)
    .map(
      (node) => `
        ${nodeTitle(node)}

        ${Object.entries(node.outputs)
          .map(([_, variable]) =>
            compileVariable({ ...variable, value: undefined })
          )
          .join("")}

        {
          /* Varying References */
          ${Object.entries(node.varyings)
            .map(([name, variable]) =>
              compileVariable({ ...variable, name, value: variable.name })
            )
            .join("")}

          /* Inputs */
          ${Object.entries(node.inputs)
            .map(([name, variable]) => compileVariable({ ...variable, name }))
            .join("")}

          /* Outputs */
          ${Object.entries(node.outputs)
            .map(([name, variable]) => compileVariable({ ...variable, name }))
            .join("")}

          /* Code */
          ${node[program].body ?? ""}

          /* Update globals */
          ${Object.entries(node.outputs)
            .map(([name, variable]) => `${variable.name} = ${name};`)
            .join("\n")}
          ${
            program === "vertex"
              ? Object.entries(node.varyings)
                  .map(([name, variable]) => `${variable.name} = ${name};`)
                  .join("")
              : ""
          }
          }
      `
    )
    .join("\n\n\n")
}

export function compileShader(root: ShaderNode) {
  /* Build uniforms */
  const uniforms = {} as Record<string, IUniform>

  dependencies(root).forEach((node) => {
    Object.entries(node.uniforms).forEach(([name, variable]) => {
      uniforms[variable.name] = variable as IUniform
    })
  })

  /* Build callbacks */
  const callbacks = dependencies(root).map((node) => node.update)

  const update: RenderCallback = (...args) => {
    callbacks.forEach((callback) => callback?.(...args))
  }

  /* Build vertex shader program */
  const vertexShader = `
    /*** VERTEX SHADER ***/

    ${compileHeader(root, "vertex")}

    void main() {
      ${compileBody(root, "vertex")}
    }`

  /* Build fragment shader program */
  const fragmentShader = `
    /*** FRAGMENT SHADER ***/

    ${compileHeader(root, "fragment")}

    void main() {
      ${compileBody(root, "fragment")}
    }`

  return {
    vertexShader,
    fragmentShader,
    uniforms,
    update
  }
}
