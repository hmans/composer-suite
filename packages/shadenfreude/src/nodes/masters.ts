import { Color } from "three"
import { Factory, vec3 } from "../shadenfreude"
import { GeometryPositionNode, GeometryNormalNode } from "./geometry"

export const CustomShaderMaterialMasterNode = Factory(() => ({
  name: "CustomShaderMaterial Master",
  in: {
    position: vec3(GeometryPositionNode()),
    normal: vec3(GeometryNormalNode()),
    diffuseColor: vec3(new Color(1, 1, 1)),
    emissiveColor: vec3(new Color(0, 0, 0))
  },
  vertex: {
    body: `
      csm_Position = in_position;
      csm_Normal = in_normal;
    `
  },
  fragment: {
    body: `
      csm_DiffuseColor.rgb = in_diffuseColor;
      csm_Emissive.rgb = in_emissiveColor;
    `
  }
}))
