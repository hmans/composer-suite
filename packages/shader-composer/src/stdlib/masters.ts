import { Color } from "three"
import { $ } from "../expressions"
import { Input } from "../units"
import { VertexPosition } from "./geometry"
import { Master } from "./values"

export type ShaderMaterialMasterProps = {
  color?: Input<"vec3">
  alpha?: Input<"float">
  position?: Input<"vec3">
}

export const ShaderMaterialMaster = ({
  color = new Color("red"),
  alpha = 1,
  position = VertexPosition
}: ShaderMaterialMasterProps = {}) =>
  Master({
    name: "ShaderMaterial Master",

    vertex: {
      body: $`
				gl_Position = projectionMatrix * modelViewMatrix * vec4(${position}, 1.0);
			`
    },

    fragment: {
      body: $`
				gl_FragColor = vec4(${color}, ${alpha});
			`
    }
  })

export type CustomShaderMaterialMasterProps = {
  position?: Input<"vec3">
  normal?: Input<"vec3">
  diffuseColor?: Input<"vec3">
  emissiveColor?: Input<"vec3">
  fragColor?: Input<"vec3">
  alpha?: Input<"float">
  roughness?: Input<"float">
  metalness?: Input<"float">
}

export const CustomShaderMaterialMaster = ({
  position,
  normal,
  diffuseColor,
  emissiveColor,
  fragColor,
  roughness,
  metalness,
  alpha
}: CustomShaderMaterialMasterProps = {}) =>
  Master({
    name: "CustomShaderMaterial Master",

    vertex: {
      body: $`
				${position !== undefined ? $`csm_Position.xyz = ${position};` : ""}
				${normal !== undefined ? $`csm_Normal = ${normal};` : ""}
			`
    },

    fragment: {
      body: $`
  			${alpha !== undefined ? $`csm_DiffuseColor.a = ${alpha};` : ""}
				${diffuseColor !== undefined ? $`csm_DiffuseColor.rgb = ${diffuseColor};` : ""}
				${emissiveColor !== undefined ? $`csm_Emissive = ${emissiveColor};` : ""}
				${fragColor !== undefined ? $`csm_FragColor = vec4(${fragColor}, ${alpha});` : ""}

        #if defined IS_MESHSTANDARDMATERIAL || defined IS_MESHPHYSICALMATERIAL
          ${roughness !== undefined ? $`csm_Roughness = ${roughness};` : ""}
          ${metalness !== undefined ? $`csm_Metalness = ${metalness};` : ""}
        #endif
			`
    }
  })
