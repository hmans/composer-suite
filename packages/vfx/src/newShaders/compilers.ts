import { formatValue } from "../shaders"
import { IVariable, Shader, VariableQualifier, Variables } from "./types"

function compileVariable(
  qualifier: VariableQualifier,
  name: string,
  { type, value }: IVariable
) {
  const valueString = value !== undefined ? ` = ${formatValue(value)}` : ""
  return `${qualifier} ${type} ${name}${valueString};`
}

function compileVariables(qualifier: VariableQualifier, variables: Variables) {
  return Object.entries(variables)
    .map(([name, variable]) => compileVariable(qualifier, name, variable))
    .join("\n")
}

function compileProgram(
  shader: Shader,
  program: "vertexShader" | "fragmentShader"
) {
  const {
    uniforms,
    varyings,
    [program]: { header, main }
  } = shader

  return `
    ${compileVariables("uniform", uniforms)}
    ${compileVariables("varying", varyings)}
    ${header}

    void main() {
      ${main}
    }`
}

export function compileMaterial(shader: Shader) {
  return {
    uniforms: shader.uniforms,
    vertexShader: compileProgram(shader, "vertexShader"),
    fragmentShader: compileProgram(shader, "fragmentShader")
  }
}
