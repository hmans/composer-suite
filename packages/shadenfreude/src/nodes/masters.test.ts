import { Color } from "three"
import { compileShader } from "../compilers"
import { CustomShaderMaterialMaster } from "./masters"

describe("CustomShaderMaterialMaster", () => {
  it("renders code for CustomShaderMaterial", () => {
    const master = CustomShaderMaterialMaster({
      diffuseColor: new Color("hotpink")
    })

    const [shader] = compileShader(master)

    expect(shader.vertexShader).toMatchSnapshot()
    expect(shader.fragmentShader).toMatchSnapshot()
  })
})
