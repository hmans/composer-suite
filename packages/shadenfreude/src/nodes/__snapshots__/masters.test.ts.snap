// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomShaderMaterialMaster renders code for CustomShaderMaterial 1`] = `
"/*** BEGIN: Position (1) ***/
varying vec3 v_vec3_Position_1;
/*** END: Position (1) ***/

/*** BEGIN: Normal (2) ***/
varying vec3 v_vec3_Normal_2;
/*** END: Normal (2) ***/

void main()
{
  /*** BEGIN: Position (1) ***/
  vec3 vec3_Position_1;
  {
    vec3 value = position;
    vec3_Position_1 = value;
    v_vec3_Position_1 = value;
  }
  /*** END: Position (1) ***/

  /*** BEGIN: Normal (2) ***/
  vec3 vec3_Normal_2;
  {
    vec3 value = normal;
    vec3_Normal_2 = value;
    v_vec3_Normal_2 = value;
  }
  /*** END: Normal (2) ***/

  /*** BEGIN: CustomShaderMaterial Master (3) ***/
  bool bool_CustomShaderMaterial_Master_3;
  {
    bool value = true;
    
      csm_Position = vec3_Position_1;
      csm_Normal = vec3_Normal_2;
    
    bool_CustomShaderMaterial_Master_3 = value;
  }
  /*** END: CustomShaderMaterial Master (3) ***/

}"
`;

exports[`CustomShaderMaterialMaster renders code for CustomShaderMaterial 2`] = `
"/*** BEGIN: Position (1) ***/
varying vec3 v_vec3_Position_1;
/*** END: Position (1) ***/

/*** BEGIN: Normal (2) ***/
varying vec3 v_vec3_Normal_2;
/*** END: Normal (2) ***/

void main()
{
  /*** BEGIN: Position (1) ***/
  vec3 vec3_Position_1;
  {
    vec3 value = v_vec3_Position_1;
    vec3_Position_1 = value;
  }
  /*** END: Position (1) ***/

  /*** BEGIN: Normal (2) ***/
  vec3 vec3_Normal_2;
  {
    vec3 value = v_vec3_Normal_2;
    vec3_Normal_2 = value;
  }
  /*** END: Normal (2) ***/

  /*** BEGIN: CustomShaderMaterial Master (3) ***/
  bool bool_CustomShaderMaterial_Master_3;
  {
    bool value = true;
    
      csm_DiffuseColor = vec4(
      vec3(
        1.0,
        0.4117647058823529,
        0.7058823529411765
      ), 1.0);
    
    bool_CustomShaderMaterial_Master_3 = value;
  }
  /*** END: CustomShaderMaterial Master (3) ***/

}"
`;
